python/lsst/pipe/tasks/make_psf_matched_warp.py

```diff
@@ -0,0 +1,215 @@
+            transform = makeWcsPairTransform(row.wcs, direct_warp.wcs)
+            warp_psf = WarpedPsf(row.getPsf(), transform)
+
+            if (destination_polygon := row.validPolygon) is None:
+                # Calculate the polygon for this detector.
+                src_polygon = Polygon(
+                    [geom.Point2D(corner) for corner in row.getBBox().getCorners()]
+                )
+                destination_polygon = src_polygon.transform(transform).intersectionSingle(
+                    geom.Box2D(direct_warp.getBBox())
+                )
+                self.log.trace("Polygon for detector=%d is calculated as %s",
+                               row["ccd"],
+                               destination_polygon
+                               )
+            else:
+                self.log.trace("Polygon for detector=%d is read from the input calexp as %s",
+                               row["ccd"],
+                               destination_polygon
+                               )
+
+            # Compute the minimum possible bounding box that overlaps the CCD.
+            # First find the intersection polygon between the per-detector warp
+            # and the warp bounding box.
+            bbox = geom.Box2I()
+            for corner in destination_polygon.getVertices():
+                bbox.include(geom.Point2I(corner))
+            bbox.clip(direct_warp.getBBox())  # Additional safeguard
+
+            self.log.trace("PSF-matching CCD %d with bbox %s", row["ccd"], bbox)
+
+            ccd_mask_array = ~(destination_polygon.createImage(bbox).array <= 0)
+
+            # Clone the subimage, set the PSF to the model and reset the planes
+            # outside the detector.
+            temp_warp = direct_warp[bbox].clone()
+            temp_warp.setPsf(warp_psf)
+            temp_warp.image.array *= ccd_mask_array
+            temp_warp.mask.array |= (~ccd_mask_array) * bit_mask
+            # We intend to divide by zero outside the detector to set the
+            # per-pixel variance values to infinity. Suppress the warning.
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", message="divide by zero", category=RuntimeWarning)
+                temp_warp.variance.array /= ccd_mask_array
+
+            temp_psf_matched = self.psfMatch.run(temp_warp, modelPsf).psfMatchedExposure
+            del temp_warp
+
+            # Set pixels outside the intersection polygon to NO_DATA.
+            temp_psf_matched.maskedImage[bbox].mask.array |= (~ccd_mask_array) * bit_mask
+
+            num_good_pixels = copyGoodPixels(
+                exposure_psf_matched.maskedImage[bbox],
+                temp_psf_matched.maskedImage[bbox],
+                bit_mask,
+            )
+
+            del temp_psf_matched
+
+            self.log.trace(
+                "Copied %d pixels from temp_warp to exposure_psf_matched", num_good_pixels
+            )
```

Comment:
Might even be worth making this INFO-level if you can put a data ID or something similar in; there'd still only be <= 4 of them per quantum.

Reply:
Do we need access to data ID explicitly to make this log (at INFO level) useful. This is how it appears to me anyway

```
INFO 2024-05-22T13:04:54.489-07:00 lsst.makePsfMatchedWarp (makePsfMatchedWarp:{instrument: 'LSSTCam-imSim', skymap: 'DC2', tract: 3828, patch: 42, visit: 214464, band: 'i', physical_filter: 'i_sim_1.4'})(make_psf_matched_warp.py:208) - Copied 3249155 pixels from temp_warp to exposure_psf_matched
```

This method knows nothing about the data ID and it won't unless we override the `runQuantum` method.

Reply:
Oh, I suppose you meant the `ccd` id of the input `calexp`.

Reply:
Yes, the visit ID isn't needed because one can get that from task start/stop logs, but he detector would make the logging a bit more useful.